internal/dataplane/parser/translators/httproute_atc.go

```diff
@@ -227,3 +229,370 @@ func matchersFromParentHTTPRoute(hostnames []string, metaAnnotations map[string]
+		HeaderNumberShiftBits = 23
+		// QueryParamNumberShiftBits makes bits 18-22 used for number of query params (max number of query params = 16)
+		QueryParamNumberShiftBits = 18
+		// bits 0-17 are used for relative order of creation timestamp, namespace/name, and internal order of rules and matches.
+		// the bits are calculated by
+	)
+
+	var priority int
+	if t.PreciseHostname {
+		priority += (1 << PreciseHostnameShiftBits)
+	}
+	priority += t.HostnameLength << HostnameLengthShiftBits
+
+	if t.PathType == gatewayv1beta1.PathMatchExact {
+		priority += (1 << ExactPathShiftBits)
+	}
+	if t.PathType == gatewayv1beta1.PathMatchRegularExpression {
+		priority += (1 << RegularExpressionPathShiftBits)
+	}
+
+	// max length of path is 1024, but path must start with /, so we use PathLength-1 to fill the bits.
+	if t.PathLength > 0 {
+		priority += ((t.PathLength - 1) << PathLengthShiftBits)
+	}
+
+	priority += (t.HeaderCount << HeaderNumberShiftBits)
+	if t.HasMethodMatch {
+		priority += (1 << MethodMatchShiftBits)
+	}
+	priority += (t.QueryParamCount << QueryParamNumberShiftBits)
+	priority += (ResourceKindBitsHTTPRoute << FromResourceKindPriorityShiftBits)
+
+	return priority
+}
+
+// AssignRoutePriorityToSplitHTTPRoutes assigns priority to ALL split HTTPRoutes
+// that are split by hostnames and matches from HTTPRoutes listed from the cache.
+// Firstly assign "fixed" bits by the following fields of the HTTPRoute:
+// hostname, path type, path length, method match, number of header matches, number of query param matches.
+// If ties exists in the first step, where multiple HTTPRoute has the same priority
+// calculated from the fields, we run a sort for the HTTPRoutes in the tie
+// and assign the bits for "relative order" according to the sorting result of these HTTPRoutes.
+func AssignRoutePriorityToSplitHTTPRoutes(
+	splitHTTPRoutes []*gatewayv1beta1.HTTPRoute,
+) []SplitHTTPRouteToKongRoutePriority {
+	priorityToHTTPRoutes := map[int][]*gatewayv1beta1.HTTPRoute{}
+
+	for _, httpRoute := range splitHTTPRoutes {
+		anns := httpRoute.Annotations
+		// skip if HTTPRoute does not contain the annotation, because this means the HTTPRoute is not a split one.
+		if anns == nil || anns[InternalRuleIndexAnnotationKey] == "" || anns[InternalMatchIndexAnnotationKey] == "" {
+			continue
+		}
+
+		priority := CalculateHTTPRoutePriorityTraits(httpRoute).EncodeToPriority()
+		priorityToHTTPRoutes[priority] = append(priorityToHTTPRoutes[priority], httpRoute)
+	}
+
+	httpRoutesToPriorities := make([]SplitHTTPRouteToKongRoutePriority, 0, len(splitHTTPRoutes))
+
+	// Bits 0-17 (18 bits) are assigned for relative order of HTTPRoutes.
+	// If multiple HTTPRoutes are assigned to the same priority in the previous step,
+	// sort them then starts with 2^18 -1 and decrease by one for each HTTPRoute;
+	// If only one HTTPRoute occupies the priority, fill the relative order bits with all 1s.
+	const RelativeOrderAssignedBits = 18
```

Comment:
Is there any specific reason not to have these consts global? It looks a bit weird to to have them local.

Reply:
Because the bit shift are only used for generating `priority` for Kong Routes from HTTPRoutes. `Ingress`es and `GRPCRoute`s also have similar definitions. If we want to make them global, we need to add `HTTPRoutePriority` prefix, but I do not know where else should we use them (maybe in tests?)